### ISSUE TITLE
AsyncConnection::EventLoop::run hangs when Redis disconnects

### ISSUE THREAD

**Issue by Taxopr** [rtbkit/rtbkit#242]
AsyncConnection::EventLoop::run thread consumes 100% CPU, this happend because of unhandled POLLERR flag in revents.
strace shows repeating records:

> poll([{fd=67, events=POLLIN}, {fd=66, events=0}], 2, 1000000) = 1 ([{fd=66, revents=POLLERR|POLLHUP}])

This means, I believe, that Redis closed connection, and main cycle of AsyncConnection spins very fast, not handling this state.
I think, we should handle POLLERR state, stop main cycle and close AsyncConnection with appropriate error.

**Comment by RAttab** [comment]
You're right, in the case of an POLLHUP the code will just spin pointlessly. I'll have to do a bit more digging to figure out what needs to happen in this scenario.

Thanks for raising the issue.

**Comment by RAttab** [comment]
On further inspection, it looks like hiredis has a disconnection detection mechanism but it's unclear how this interacts with our poller and the documentation seems a bit light on this point. In any case, I'll have to do a bit more digging. 

Also it looks as if even we proper handle the error, we won't be trying to reconnect to the redis server automatically which would probably lead to an error loop downstream. It might take a while to figure out a fix for this.

In the meantime, is this a recurrent or intermittent issue? My guess is that it only occurs when you lose your connection to redis. If so I'd recommend you focus on trying to solve that issue.

**Comment by Taxopr** [comment]
I think, trying to reconnect to redis server is higher-level feature, than Redis::AsyncConnection, so it's OK just handle disconnect without trying to reconnect internally.
Anyway, such behaviour won't make things worse than now. Now we just hang completely one of server cores.
For me this is recurrent issue, I believe, this issue occurs when we lose connection to redis. Maybe redis some why closes connection on its side. I'm investigating it.
Anyway, IMHO, this issue should be solved to improve stability of rtbkit.

**Comment by RAttab** [comment]
Makes sense. I'll try to find a hole this week to fix this issue. Note that a contribution would also be welcome.

**Comment by RAttab** [comment]
Little update. I believe we have a working fix on our branch written by @oktal. All that's left is to upstream it which should be done soon-ish.

**Comment by Taxopr** [comment]
What is the name of this branch?

**Comment by RAttab** [comment]
You can find the PR here: https://github.com/datacratic/soa/pull/183

**Comment by RAttab** [comment]
The PR is on the datacratic/soa which is our internal fork of soa that we use internally to avoid needlessly destabilizing rtbkit. 

The simplest way for you to try out the fix without pulling a bunch of random commits would be to either download the raw patch from github. Note that we use git subtree to upstream changes into the rtbkit repo but that's a bit more involved and you'd bring a bunch of other changes as well. Note that we will apply the fix to rtbkit so you could also just wait for that.

**Comment by RAttab** [comment]
So we created an upstream branch for the latest batches of changes (https://github.com/rtbkit/rtbkit/pull/244). This means that you can access the redis fix easily by cherry-picking the following commits: https://github.com/datacratic/rtbkit/commit/07d88ad9b3fc967af0c6342b059e1690b5e9ab9c and https://github.com/datacratic/rtbkit/commit/1f452f76d3f444c3efbeda461915c5ca22230b4d. The upstream PR will be merged sometime this weeks.

Let me if it resolves your issue.
